File: CS/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraPivotGrid;
using System.Reflection;
using DevExpress.XtraPivotGrid.Data;
using DevExpress.XtraPivotGrid.ViewInfo;
using DevExpress.Utils.Drawing;

namespace Q205054 {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e) {
            // TODO: This line of code loads data into the 'nwindDataSet.ProductReports' table. You can move, or remove it, as needed.
            this.productReportsTableAdapter.Fill(this.nwindDataSet.ProductReports);

        }

        private void pivotGridControl1_MouseClick(object sender, MouseEventArgs e) {
            PivotGridHitInfo hInfo = pivotGridControl1.CalcHitInfo(e.Location);
            if(hInfo.HitTest == PivotGridHitTest.Value)
                HandleValueMouseClick(hInfo.ValueInfo);
        }

        private void HandleValueMouseClick(PivotFieldValueEventArgs e) {
            PivotGridField[] higherFields = e.GetHigherLevelFields();
            object[] higherValues = new object[higherFields.Length];
            for(int i = 0; i < higherFields.Length; i++)
                higherValues[i] = e.GetHigherLevelFieldValue(higherFields[i]);

            pivotGridControl1.BeginUpdate();
            PivotArea otherArea = GetOtherArea(e.IsColumn);
            List<PivotGridField> otherFields = pivotGridControl1.GetFieldsByArea(otherArea);
            PivotSortOrder? sortOrder = GetSummarySortOrder(e.Item);
            for(int i = 0; i < otherFields.Count; i++) {
                if(!sortOrder.HasValue || sortOrder.Value == PivotSortOrder.Descending)
                    otherFields[i].SortOrder = PivotSortOrder.Ascending;
                else
                    otherFields[i].SortOrder = PivotSortOrder.Descending;
                otherFields[i].SortBySummaryInfo.Field = e.DataField;
                otherFields[i].SortBySummaryInfo.Conditions.Clear();
                for(int j = 0; j < higherFields.Length; j++) {
                    otherFields[i].SortBySummaryInfo.Conditions.Add(new PivotGridFieldSortCondition(higherFields[j], higherValues[j]));
                }
                if(e.Field != null && e.Field.Area != PivotArea.DataArea)
                    otherFields[i].SortBySummaryInfo.Conditions.Add(new PivotGridFieldSortCondition(e.Field, e.Value));                
            }
            pivotGridControl1.EndUpdate();
        }

        private PivotArea GetOtherArea(bool isColumn) {
            return isColumn ? PivotArea.RowArea : PivotArea.ColumnArea;
        }

        PropertyInfo viewInfoPI = null;
        PivotFieldValueItem GetItem(PivotCustomDrawFieldValueEventArgs e) {
            if(viewInfoPI == null)
                viewInfoPI = e.GetType().GetProperty("FieldCellViewInfo", BindingFlags.Instance | BindingFlags.NonPublic);
            PivotFieldsAreaCellViewInfo viewInfo = (PivotFieldsAreaCellViewInfo)viewInfoPI.GetValue(e, null);
            return viewInfo.Item;
        }

        private PivotSortOrder? GetSummarySortOrder(PivotFieldValueItem valueItem) {
            if(!valueItem.IsLastFieldLevel)
                return null;
            List<PivotGridFieldPair> pairs = valueItem.Data.VisualItems.GetSortedBySummaryFields(valueItem.IsColumn, valueItem.Index);
            if(pairs == null)
                return null;
            PivotSortOrder? sortOrder = null;
            foreach(PivotGridFieldPair pair in pairs) {
                if(pair.DataFieldItem != valueItem.DataField) continue;
                if(sortOrder != null) {
                    if(sortOrder.Value != pair.Field.SortOrder) {
                        sortOrder = null;
                        break;
                    }
                }
                sortOrder = pair.Field.SortOrder;
            }
            return sortOrder;
        }

        private void pivotGridControl1_CustomDrawFieldValue(object sender, PivotCustomDrawFieldValueEventArgs e) {
            PivotFieldValueItem valueItem = GetItem(e);            
            PivotSortOrder? sortOrder = GetSummarySortOrder(valueItem);           
            if(sortOrder == null) 
                return; // proceed to standard drawing
            PivotGridViewInfoData data = (PivotGridViewInfoData)valueItem.Data;
            data.ActiveLookAndFeel.Painter.Header.DrawObject(e.Info);


            SortedShapeObjectInfoArgs sortInfo = new SortedShapeObjectInfoArgs();
            sortInfo.Ascending = sortOrder == PivotSortOrder.Ascending;
            sortInfo.Graphics = e.Graphics;
            Rectangle sortBounds = data.ActiveLookAndFeel.Painter.SortedShape.CalcObjectMinBounds(sortInfo);
            sortBounds.X = e.Info.CaptionRect.Right + 2;
            sortBounds.Y = e.Info.CaptionRect.Y + (int)Math.Round((double)(e.Info.CaptionRect.Height - sortBounds.Height) / 2);
            sortInfo.Bounds = sortBounds;
            data.ActiveLookAndFeel.Painter.SortedShape.DrawObject(sortInfo);
            e.Handled = true;
        }
    }
}

File: VB/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraPivotGrid
Imports System.Reflection
Imports DevExpress.XtraPivotGrid.Data
Imports DevExpress.XtraPivotGrid.ViewInfo
Imports DevExpress.Utils.Drawing

Namespace Q205054
	Partial Public Class Form1
		Inherits Form

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
			' TODO: This line of code loads data into the 'nwindDataSet.ProductReports' table. You can move, or remove it, as needed.
			Me.productReportsTableAdapter.Fill(Me.nwindDataSet.ProductReports)

		End Sub

		Private Sub pivotGridControl1_MouseClick(ByVal sender As Object, ByVal e As MouseEventArgs) Handles pivotGridControl1.MouseClick
			Dim hInfo As PivotGridHitInfo = pivotGridControl1.CalcHitInfo(e.Location)
			If hInfo.HitTest = PivotGridHitTest.Value Then
				HandleValueMouseClick(hInfo.ValueInfo)
			End If
		End Sub

		Private Sub HandleValueMouseClick(ByVal e As PivotFieldValueEventArgs)
			Dim higherFields() As PivotGridField = e.GetHigherLevelFields()
			Dim higherValues(higherFields.Length - 1) As Object
			For i As Integer = 0 To higherFields.Length - 1
				higherValues(i) = e.GetHigherLevelFieldValue(higherFields(i))
			Next i

			pivotGridControl1.BeginUpdate()
			Dim otherArea As PivotArea = GetOtherArea(e.IsColumn)
			Dim otherFields As List(Of PivotGridField) = pivotGridControl1.GetFieldsByArea(otherArea)
			Dim sortOrder? As PivotSortOrder = GetSummarySortOrder(e.Item)
			For i As Integer = 0 To otherFields.Count - 1
				If Not sortOrder.HasValue OrElse sortOrder.Value = PivotSortOrder.Descending Then
					otherFields(i).SortOrder = PivotSortOrder.Ascending
				Else
					otherFields(i).SortOrder = PivotSortOrder.Descending
				End If
				otherFields(i).SortBySummaryInfo.Field = e.DataField
				otherFields(i).SortBySummaryInfo.Conditions.Clear()
				For j As Integer = 0 To higherFields.Length - 1
					otherFields(i).SortBySummaryInfo.Conditions.Add(New PivotGridFieldSortCondition(higherFields(j), higherValues(j)))
				Next j
				If e.Field IsNot Nothing AndAlso e.Field.Area <> PivotArea.DataArea Then
					otherFields(i).SortBySummaryInfo.Conditions.Add(New PivotGridFieldSortCondition(e.Field, e.Value))
				End If
			Next i
			pivotGridControl1.EndUpdate()
		End Sub

		Private Function GetOtherArea(ByVal isColumn As Boolean) As PivotArea
			Return If(isColumn, PivotArea.RowArea, PivotArea.ColumnArea)
		End Function

		Private viewInfoPI As PropertyInfo = Nothing
		Private Function GetItem(ByVal e As PivotCustomDrawFieldValueEventArgs) As PivotFieldValueItem
			If viewInfoPI Is Nothing Then
				viewInfoPI = e.GetType().GetProperty("FieldCellViewInfo", BindingFlags.Instance Or BindingFlags.NonPublic)
			End If
			Dim viewInfo As PivotFieldsAreaCellViewInfo = DirectCast(viewInfoPI.GetValue(e, Nothing), PivotFieldsAreaCellViewInfo)
			Return viewInfo.Item
		End Function

		Private Function GetSummarySortOrder(ByVal valueItem As PivotFieldValueItem) As PivotSortOrder?
			If Not valueItem.IsLastFieldLevel Then
				Return Nothing
			End If
			Dim pairs As List(Of PivotGridFieldPair) = valueItem.SortedBySummaryFields
			If pairs Is Nothing Then
				Return Nothing
			End If
			Dim sortOrder? As PivotSortOrder = Nothing
			For Each pair As PivotGridFieldPair In pairs
				If pair.DataFieldItem IsNot valueItem.DataField Then
					Continue For
				End If
				If sortOrder IsNot Nothing Then
					If sortOrder.Value <> pair.FieldItem.SortOrder Then
						sortOrder = Nothing
						Exit For
					End If
				End If
				sortOrder = pair.FieldItem.SortOrder
			Next pair
			Return sortOrder
		End Function

		Private Sub pivotGridControl1_CustomDrawFieldValue(ByVal sender As Object, ByVal e As PivotCustomDrawFieldValueEventArgs) Handles pivotGridControl1.CustomDrawFieldValue
			Dim valueItem As PivotFieldValueItem = GetItem(e)
			Dim sortOrder? As PivotSortOrder = GetSummarySortOrder(valueItem)
			If sortOrder Is Nothing Then
				Return ' proceed to standard drawing
			End If
            Dim data As PivotGridViewInfoData = CType(pivotGridControl1, IPivotGridViewInfoDataOwner).DataViewInfo
            e.Info.InnerElements.Clear()
            data.ActiveLookAndFeel.Painter.Header.DrawObject(e.Info)


			Dim sortInfo As New SortedShapeObjectInfoArgs()
            sortInfo.Ascending = CBool(sortOrder = PivotSortOrder.Ascending)
            sortInfo.Graphics = e.Graphics
			Dim sortBounds As Rectangle = data.ActiveLookAndFeel.Painter.SortedShape.CalcObjectMinBounds(sortInfo)
			sortBounds.X = e.Info.CaptionRect.Right + 2
			sortBounds.Y = e.Info.CaptionRect.Y + CInt(Fix(Math.Round(CDbl(e.Info.CaptionRect.Height - sortBounds.Height) / 2)))
			sortInfo.Bounds = sortBounds
			data.ActiveLookAndFeel.Painter.SortedShape.DrawObject(sortInfo)
			e.Handled = True
		End Sub
	End Class
End Namespace